File: sources/java/_11_parallel_sum.java
// task 11 parallel_sum — expected output: 7500000075000000
// build: javac _11_parallel_sum.java    run: java _11_parallel_sum
// build (graalvm native-image): native-image -O2 _11_parallel_sum    run: ./_11_parallel_sum
// Four real OS threads (java.lang.Thread) over fixed ranges; the partial sums are exact in long.

public class _11_parallel_sum {
    private static final long SPAN = 25000000L;

    private static long work(long t) {
        long start = t * SPAN;
        long end = start + SPAN;
        long acc = 0;
        for (long i = start; i < end; i++) {
            switch ((int) (i % 4)) {
                case 0:
                    acc += 1;
                    break;
                case 1:
                    acc += i;
                    break;
                case 2:
                    acc += 2 * i;
                    break;
                case 3:
                    acc += 3 * i;
                    break;
            }
        }
        return acc;
    }

    public static void main(String[] args) throws InterruptedException {
        long[] results = new long[4];
        Thread[] threads = new Thread[4];
        for (int t = 0; t < 4; t++) {
            final int id = t;
            threads[t] = new Thread(() -> results[id] = work(id));
        }
        for (Thread th : threads) {
            th.start();
        }
        for (Thread th : threads) {
            th.join();
        }
        long total = 0;
        for (int t = 0; t < 4; t++) {
            total += results[t];
        }
        System.out.println(total);
    }
}

File: sources/scala/11_parallel_sum.scala
// task 11 parallel_sum — expected output: 7500000075000000
// build: scalac -release 17 11_parallel_sum.scala    run: scala Main

object Main {
  private def work(t: Int, results: Array[Long]): Unit = {
    val start = t.toLong * 25000000L
    val end = start + 25000000L
    var acc = 0L
    var i = start
    while (i < end) {
      (i % 4L).toInt match {
        case 0 => acc += 1L
        case 1 => acc += i
        case 2 => acc += 2L * i
        case 3 => acc += 3L * i
      }
      i += 1L
    }
    results(t) = acc
  }

  def main(args: Array[String]): Unit = {
    val results = new Array[Long](4)
    val threads = new Array[Thread](4)
    var t = 0
    while (t < 4) {
      val idx = t
      threads(t) = new Thread(new Runnable {
        def run(): Unit = work(idx, results)
      })
      t += 1
    }
    t = 0
    while (t < 4) {
      threads(t).start()
      t += 1
    }
    t = 0
    while (t < 4) {
      threads(t).join()
      t += 1
    }
    var total = 0L
    t = 0
    while (t < 4) {
      total += results(t)
      t += 1
    }
    println(total)
  }
}
